Control transform (#1157)
PennyLaneAI/pennylane: .github/CHANGELOG.md

<h3>New features since last release</h3>

* Adds a new transform `qml.ctrl` that adds control wires to subroutines.
  [(#1157)](https://github.com/PennyLaneAI/pennylane/pull/1157)

  Here's a simple usage example:

  ```python
  def my_anzats(params):
     qml.RX(params[0], wires=0)
     qml.RZ(params[1], wires=1)

  # Create a new method that applies `my_anzats`
  # controlled by the "2" wire.
  my_anzats2 = ctrl(my_anzats, control=2)

  @qml.qnode(...)
  def circuit(params):
      my_anzats2(params)
      return qml.state()
  ```

  The above `circuit` would be equivalent to:

  ```python
  @qml.qnode(...)
  def circuit(params):
      qml.CRX(params[0], wires=[2, 0])
      qml.CRZ(params[1], wires=[2, 1])
      return qml.state()
  ```

>   The `qml.ctrl` transform is especially useful for when you want to repeatedly apply an
>   operation but controlled by different qubits. A very famous example that can benefit from 

Reviewer: ```suggestion
  The `qml.ctrl` transform is especially useful to repeatedly apply an
  operation which is controlled by different qubits in each repetition. A very famous example that can benefit from 
```

>   The `qml.ctrl` transform is especially useful for when you want to repeatedly apply an
>   operation but controlled by different qubits. A very famous example that can benefit from 
>   this is Shor's algorithm.

Reviewer: ```suggestion
  The `qml.ctrl` transform is especially useful to repeatedly apply an
  operation which is controlled by different qubits in each repetition. A famous example is Shor's algorithm.
```

  ```python
  def modmul(a, mod, wires):
      # Some complex set of gates that implements modular multiplcation.
      # qml.CNOT(...); qml.Toffoli(...); ... 
 
  @qml.qnode(...)

>   def shors(a, mod, scratch_wires, qft_wires):

Reviewer: ```suggestion
  def shor(a, mod, scratch_wires, qft_wires):
```
?

<h3>New features since last release</h3>

* Added `qml.ctrl`, a transformation that adds control wires to subroutines.
  [(#1157)](https://github.com/PennyLaneAI/pennylane/pull/1157)

  Here's a simple example

  ```python

>   def my_anzats(params):

Reviewer: That sounds a bit like a Transylvanian version :)

  @qml.qnode(...)
  def circuit(params):
      my_anzats2(params)
      return qml.state()
  ```

  The above `circuit` would be equivalent to:

  ```python
  @qml.qnode(...)
  def circuit(params):
      qml.CRX(params[0], wires=[2, 0])
      qml.CRZ(params[1], wires=[2, 1])
      return qml.state()
  ```

  The `qml.ctrl` transform is especially useful for when you want to repeatedly apply an
  operation but controlled by different qubits. A very famous example that can benefit from 
  this is Shor's algorithm.

  ```python
  def modmul(a, mod, wires):
      # Some complex set of gates that implements modular multiplcation.
      # qml.CNOT(...); qml.Toffoli(...); ...
      ...
 
 
  @qml.qnode(...)
  def shors(a, mod, scratch_wires, qft_wires):
      for i, wire in enumerate(qft_wires):
          qml.Hadamard(wire)
     
          # Create the controlled modular multiplication 
          # subroutine based on the control wire.
          cmodmul = qml.ctrl(modmul, control=wire)

          # Execute the controlled modular multiplication.
          cmodmul(a ** i, mod, scratch_wires)
 
      qml.adjoint(qml.QFT)(qft_wires)
      return qml.sample()
  ```

>   In the future, devices will be able to expliot the sparsity of controlled operations to 

Reviewer: Is it "in future"?
